### PR TITLE
Auto responder feature

### PR DESCRIPTION
**Pull request by csduarte**
#### Summary
Enable AutoResponder and add a custom message
When status is set to Out Of Office display a confirmation when user changes status manually, because changing status will disable Auto Responder.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: [iPhone 6 simulator, Nexus 5X emulator] 

#### Setting Auto Responder
screen shot 2018-03-08 at 1 36 45 pm
screen shot 2018-03-08 at 1 36 27 pm

#### Disabling Auto Responder
screen shot 2018-03-08 at 1 37 53 pm

#### Disabling Auto Responder in action
ooo_in_action

via @miguelespinoza

### CODE REVIEW

File: app/screens/settings/notification_settings/notification_settings.js [hunk 1]
```diff
@@ -375,6 +433,16 @@ class NotificationSettings extends PureComponent {
...
                         iconType='ion'
                         onPress={() => this.handlePress(this.goToNotificationSettingsEmail)}
+                        separator={true}
+                        showArrow={showArrow}
+                        theme={theme}
+                    />
+                    <SettingsItem
+                        defaultMessage='Out Of Office: Auto Responder'
+                        i18nId='mobile.notification_settings.ooo_auto_responder'
+                        iconName='airplanemode-active'
```
Comment (cpanato) [inline app/screens/settings/notification_settings/notification_settings.js]: maybe can use this icon: https://www.materialui.co/icon/beach-access 😄

**Comment by enahum** [comment]
Question before doing any review, why would you show an Alert to get out of the OOF status when I'm explicitly doing so?

**Comment by koxen** [comment]
just curious, on the server side is OOO treated as away? offline? do not disturb?
has push notifications code on server side been updated to handle this?

**Comment by koxen** [comment]
I can't see any PR on the mattermost-server side. Shouldn't these messages be sent by the server so they work if your phone is offline?

**Comment by miguelespinoza** [comment]
@enahum here's the reason we display an Alert to get out of OOF.

When setting the Auto Responder the status of the user is set to Out Of Office in the server.
Say a user is in vacation and regularly checks in with the team.  

Notifications are silent when OOO status is enabled, treated the same as DnD.

We discovered a minor nuisance on web, a user has the ability to set their status back to online/away/dnd/offline and disabling the AutoResponder settings.

Note the rules for the AutoResponder in the server PR:
https://github.com/mattermost/mattermost-server/pull/8386
> * When enabled, it allows you to set a custom message that will be sent to anyone who DMs you.
> * This sets a new status called **Out Of Office**, this will disable notifications
> * If user decides to change status manually, they will receive a confirmation on both (web and mobile) that they'll be disabling the Auto Responder.
> * In the api, the AutoResponder get's disabled and updates all clients accordingly with a websocket event.


As a user who is Out Of Office (in vacation), the last thing you want is to have to go back and forth enabling/disabling the Auto Responder.  Therefore, we give them a confirmation popup on both web and mobile that informs them the implications of their actions.

**Comment by miguelespinoza** [comment]
Yes my bad @koxen , I forgot to link the server and webapp PRs:


* server: https://github.com/mattermost/mattermost-server/pull/8386
* webapp: https://github.com/mattermost/mattermost-webapp/pull/890

**Comment by enahum** [comment]
@michaelgamble thoughts on the UX?

**Comment by esethna** [comment]
FYI all, we are in the process of reviewing this and the webapp PRs with the UX team, will circle back with feedback soon.

**Comment by esethna** [comment]
Thanks @miguelespinoza, really appreciate your work on this feature!

UX team has had a chance to discuss and we have some feedback to share. We're keen on getting this feature in and working with you and your team on any updates and design changes. Open to discussion on any points here.

Sounds like the end goal is **automatically respond to users who are attempting to reach out-of-office (OOF) team members.**

Our main UX concern regarding the PR as implemented is mixing the OOF feature with the status dropdown, as it may not be necessary to accomplish the goal. The intention of the online status is to express your visibility to others on the team and configure your level of notifications.  If OOF is built in as part of the status dropdown then it eliminates the ability to set other statuses that can alter notifications. We think we should separate the ideas of setting an auto-response and setting an online status, because, ultimately, we'd like to leave the door open to handle "Out-of-Office" and other temporary statuses using custom statuses (setting emoji and custom text in a status). We are worried that if we introduce OOF as part of the status dropdown now, that we are closing that door in the future.

So if we think about removing the OOF status and treating the auto responder as a stand alone feature, there are a couple of new design considerations introduced:
- What is the new mechanism in which the OOF user is made aware that the autoresponder is on? 
- What is the new mechanism to quickly be able to turn off the auto responder (other than going into account settings)?

To address the above we were thinking about a couple design options:

1. Reminder on app refresh with an option to turn off
2. Ephemeral system message reminder (with link to turn it off) if the OOF user posts a message.    
3. Adding an off date/time to the auto responder account setting, so it would automatically turn off, similar to what Outlook offers?
4. App banner with a link to turn it off (similar to Outlook).

Our UX team is working on a few design options but we would love to hear your feedback and if you have any other ideas on how to move forward with this.

Also, adding @michaelgamble and @asaadmahmood who are helping with the designs and review.

**Comment by csduarte** [comment]
@esethna  I'm not against the change, but I'm not sure I understand the reasoning. Out of Office is a fairly standard status and don't really block custom status. I think our need/implementation may be too narrowly scoped. If we're trying to widen/change the scope, please schedule a meeting to discuss. I think there is too much to cover for this thread.

**Comment by miguelespinoza** [comment]
@esethna 
Could you please schedule a meeting to discuss this.

There are 2 goals with this feature.
1. automatically respond to users who are attempting to reach out-of-office (OOF) team members.
2. out-of-office users accomplish; enabling an automated custom message and disabling notifications.

In order to accomplish both goals, we introduced "Out Of Office" like Chris mentioned a fairly standard status.

In short, the OOF status allows us to work with an existing concept that users are familiar with.  We can leverage this to tightly integrate the AutoResponder feature with the existing system.

One thing to note,
The OOF status is only ever visible/enabled when the AutoResponder is enabled.

When the AutoResponder is disabled, the OOF status is not visible through the dropdown.

**Comment by esethna** [comment]
Hi @miguelespinoza,

Following our Zoom conversation today we have discussed internally with UX. Our proposal for moving forward with this feature in v1 and accomplishing your requirements is:

1. Let's please add an experimental config flag to show this feature in account settings:`ExperimentalEnableAutomaticReplies`
2. In the account setting, let's please make a few text adjustments:
    a. Setting title: "Auto Responder" > "Automatic Direct Message Replies"
    b. Help text: "Set a custom message that will be automatically sent in response to Direct Messages. Mentions in Public and Private Channels will not trigger the automated reply. Enabling Automatic Replies disables email and push notifications."
2. As a first iteration on this feature, we think it's safest to remove the "Out of Office" status that is enabled when the responder is turned on - so the user will disable from account settings.
4. We want to verify that the system message posted in the DM channel is visible to both users, since it would be important that the OOO user also see's the system message so they are aware the responder is enabled.
5. For the system message reply, can we please use the same styling and classes as the "BOT" tag, to ensure the label is properly themed. Can we also change the text to "Automatic Reply"

**Comment by miguelespinoza** [comment]
Alrighty, the changes @esethna requested are in 😄 

Now for the important question, which should we use for the Auto Responder feauture?
airplane or umbrella icon.  thanks to @cpanato and his input.  I like the umbrella icon also 😄 

#### Umbrella icon
screen shot 2018-03-30 at 12 39 18 pm

#### Airplane icon
screen shot 2018-03-30 at 12 38 39 pm

What does MM think? @esethna thoughts?

**Comment by esethna** [comment]
@miguelespinoza, hoping to get this in for v1.9, a couple tweaks if we can:

1. Can we update the text on the status change screen when manually choosing a status:

```
Disable "Out of Office"?
Would you like to switch your status to "XXX" and disable Automatic Replies?
```

2. Can we update the setting title in Notification settings to "Automatic Direct Message Replies" and update the help text to be same as webapp:

```
Set a custom message that will be automatically sent in response to Direct Messages. Mentions in Public and Private Channels will not trigger the automated reply. Enabling Automatic Replies sets your status to Out of Office and disables email and push notifications.
```

3. I'll check with @michaelgamble if he has thoughts on the icon. I'm good with the Umbrella for now.

**Comment by michaelgamble** [comment]
Regarding the autoresponder Icon - I would avoid the airplane as to me it too strongly implies "Airplane Mode" for phones - I'm more in favor of the beach umbrella.

A second option might be a robot head with a speech bubble or something visual to imply the automated side of it.

**Comment by miguelespinoza** [comment]
@esethna the changes have been made:

screen shot 2018-05-29 at 12 24 25 pm

screen shot 2018-05-29 at 12 23 30 pm

Just waiting on a recommendation for an icon.
@esethna, @michaelgamble let me know what icon will work for this feature

File: app/screens/settings/notification_settings_auto_responder/index.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline app/screens/settings/notification_settings_auto_responder/index.js]: 2015

File: app/screens/settings/notification_settings_auto_responder/notification_settings_auto_responder.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline app/screens/settings/notification_settings_auto_responder/notification_settings_auto_responder.js]: same here

File: app/utils/status.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline app/utils/status.js]: and here

File: app/components/post_textbox/post_textbox.js [hunk 5]
```diff
@@ -5,13 +5,14 @@ import React, {PureComponent} from 'react';
...
 import {intlShape} from 'react-intl';
-import {RequestStatus} from 'mattermost-redux/constants';
+import {General, RequestStatus} from 'mattermost-redux/constants';
 import EventEmitter from 'mattermost-redux/utils/event_emitter';
 
 import AttachmentButton from 'app/components/attachment_button';
 import Autocomplete from 'app/components/autocomplete';
 import FileUploadPreview from 'app/components/file_upload_preview';
 import {INITIAL_HEIGHT, INSERT_TO_COMMENT, INSERT_TO_DRAFT, IS_REACTION_REGEX, MAX_CONTENT_HEIGHT, MAX_FILE_COUNT} from 'app/constants/post_textbox';
+import {confirmOutOfOfficeDisabled} from 'app/utils/status';
```
Comment (enahum) [inline app/components/post_textbox/post_textbox.js]: thinking that this one should be an inline require, thoughts?

File: app/components/settings_drawer/settings_drawer.js [hunk 6]
```diff
@@ -21,6 +21,7 @@ import SafeAreaView from 'app/components/safe_area_view';
 import Drawer from 'app/components/drawer';
 import UserStatus from 'app/components/user_status';
 import {NavigationTypes} from 'app/constants';
+import {confirmOutOfOfficeDisabled} from 'app/utils/status';
```
Comment (enahum) [inline app/components/settings_drawer/settings_drawer.js]: inline require?

File: app/screens/settings/notification_settings/notification_settings.js [hunk 7]
```diff
@@ -188,6 +213,40 @@ class NotificationSettings extends PureComponent {
         }
     };
 
+    /**
+     * shouldSaveAutoResponder
+     *
+     * Necessary in order to properly update the notifyProps when
+     * enabling/disabling the Auto Responder.
+     *
+     * Reason being, on mobile when the AutoResponder is disabled on the server
```
Comment (hmhealey) [inline app/screens/settings/notification_settings/notification_settings.js]: Is this something missing from the mobile app's websocket handling? Should we file a ticket to investigate this further?

File: app/utils/status.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+import {
+    Alert,
+} from 'react-native';
+import {toTitleCase} from 'app/utils/general';
+
+export function confirmOutOfOfficeDisabled(intl, status, updateStatus) {
+    const userStatusId = 'modal.manual_status.auto_responder.message_' + status;
+
+    let normalizeStatus = status;
```
Comment (hmhealey) [inline app/utils/status.js]: These status names should be translated

File: assets/base/i18n/en.json [hunk 9]
```diff
@@ -2757,6 +2768,7 @@
   "status_dropdown.set_away": "Away",
   "status_dropdown.set_dnd": "Do Not Disturb",
   "status_dropdown.set_dnd.extra": "Disables Desktop and Push Notifications",
+  "status_dropdown.set_ooo": "Out Of Office",
```
Comment (hmhealey) [inline assets/base/i18n/en.json]: This string already exists below. Also, we shouldn't change non-mobile translation strings in this repo since those will get overwritten by the translation server

File: app/screens/settings/notification_settings_auto_responder/notification_settings_auto_responder.js [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    onAutoResponseToggle = (active) => {
+        if (active) {
+            this.setState({auto_responder_active: 'true'});
+            return;
+        }
+        this.setState({auto_responder_active: 'false'});
+    };
+
+    onAutoResponseChangeText = (message) => {
+        return this.setState({auto_responder_message: message});
```
Comment (hmhealey) [inline app/screens/settings/notification_settings_auto_responder/notification_settings_auto_responder.js]: There shouldn't be a return statement here

File: assets/base/i18n/en.json (hunk 9 above)
Comment (miguelespinoza) [reply assets/base/i18n/en.json]: Yeh, my bad.  Must have been from moving back and forth between mobile and webapp implementation

File: app/screens/settings/notification_settings/notification_settings.js (hunk 7 above)
Comment (miguelespinoza) [reply app/screens/settings/notification_settings/notification_settings.js]: Yes, would be great to investigate what's missing on the mobile side

File: app/components/settings_drawer/settings_drawer.js (hunk 6 above)
Comment (miguelespinoza) [reply app/components/settings_drawer/settings_drawer.js]: @enahum as for the OOF work being included as an inline-require.  I'm not sure it's necessary.

It's not crucial to the cold start experience

**Comment by miguelespinoza** [comment]
@enahum and @hmhealey made all necessary changes

**Comment by miguelespinoza** [comment]
@enahum this PR is ready to go 😄

File: app/screens/settings/notification_settings/notification_settings.js (hunk 7 above)
Comment (hmhealey) [reply app/screens/settings/notification_settings/notification_settings.js]: @miguelespinoza I know it's been a while, so what is the actual behaviour that is a bug here? Is the mobile app not receiving the updated status or are the updated notify_props not being received?
